### ISSUE TITLE
[multi-classification] How to handle labels in test set that are unseen in training set?

### ISSUE THREAD

**Issue by JivanRoquet** [microsoft/LightGBM#4626]
**TL;DR:**

When using `fit()` with a test set, how to handle labels in test set that are unseen in training set (apart from removing these rows from the test set)?

**Longer explanation**

When fitting the `lightgbm.LGBMClassifier` model from Python API, I get a `ValueError: y contains previously unseen labels` when some labels are present in the eval set that are not in the training set.

For instance, training set:

feature_a | target_label
------------ | -------------
1 | 'foo'
2 | 'bar'
1 | 'foo'
2 | 'bar'
3 | 'baz'
1 | 'bar'

And test set:

feature_a | target_label
------------ | -------------
2 | 'foo'
3 | 'qux'

The following code:

```python
model = LGBMClassifier(objective='multiclass')
model.fit(X_train, y_train, eval_set=(X_test, y_test), eval_metric='multiclass')
```

Will lead to this error:

```
ValueError: y contains previously unseen labels: 'qux'
```

I understand technically why this happens (the label encoder, fitted on the training set, doesn't know about `cux`) but I don't understand the rationale behind it — shouldn't it simply consider that a label that is unknown represents a misclassification here? For instance, the model predicted `bar` for this row, and since it encounters something it hasn't seen before, it would simply say "nope, wrong prediction" and call it a day?

Beyond this, I don't understand how to deal with this problem. I don't want to remove the `cux` line from my testing set, because that would artificially bump the testing accuracy. I want the model to consider a misclassification in this example.

Any idea on how to proceed?

One possibility I had in mind was to 1. predict the (encoded) labels then 2. decode them and 3. compare with the (plain, non-encoded) test labels (and hence detect that `'cuz' != 'bar'`. But I don't know if this is possible, it seems that `LGBMClassifier.fit()` absolutely wants to encode the test target labels first, for some reason.

**Comment by JivanRoquet** [comment]
After a bit of digging, it appears that this is an issue only for the `fit()` method, not for the `predict()`.

I ended up removing the rows with a target label unseen in the training set, and this serves as a good enough eval set during training.
